### PR TITLE
Try Adding Support for XForwardedPort

### PR DESCRIPTION
**Pull request by say25**
# Support for XForwardedPort

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description

Add support for XForwardedPort - https://github.com/dotnet/aspnetcore/issues/5996

Fixes #{bug number} (in this specific format)

### CODE REVIEW

**Comment by say25** [comment]
Please note, I think this is one of my first commits to dotnet/aspnetcore and I'm not really an expert in this area but we have a need for it in my org and I saw a request for a PR - https://github.com/dotnet/aspnetcore/issues/5996#issuecomment-320971147

**Comment by Tratcher** [comment]
Welcome.

There are some compiler errors to address:
https://github.com/dotnet/aspnetcore/pull/50338/checks?check_run_id=16220918512

File: src/Middleware/HttpOverrides/src/ForwardedHeaders.cs [hunk 1]
```diff
@@ -30,7 +30,11 @@ public enum ForwardedHeaders
     /// </summary>
     XForwardedPrefix = 1 << 3,
     /// <summary>
+    /// Process X-Forwarded-Port, which identifies the port used by the client.
+    /// </summary>
+    XForwardedPort = 1 << 3,
```
Comment (adityamandaleeka) [inline src/Middleware/HttpOverrides/src/ForwardedHeaders.cs:35]: We want these flags to be unique (right now it's the same value as the prior one). More info about Flags here: https://learn.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=net-7.0

File: src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs [hunk 2]
```diff
@@ -167,14 +167,32 @@ public void ApplyForwarders(HttpContext context)
             entryCount = Math.Max(forwardedHost.Length, entryCount);
         }
 
+        if (_options.ForwardedHeaders.HasFlag(ForwardedHeaders.XForwardedPort))
+        {
+            checkPort = true;
+            forwardedPort = requestHeaders.GetCommaSeparatedValues(_options.ForwardedPortHeaderName);
+            if (_options.RequireHeaderSymmetry
+                && ((checkFor && forwardedFor!.Length != forwardedPrefix.Length)
```
Comment (adityamandaleeka) [inline src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs:175]: The intent here is to check that the For ones match the length of the Port ones you just got (same below).

File: src/Middleware/HttpOverrides/src/ForwardedHeaders.cs (hunk 1 above)
Comment (say25) [reply src/Middleware/HttpOverrides/src/ForwardedHeaders.cs:35]:
```suggestion
    XForwardedPort = 1 << 4,
```
Oops copy pasta

**Comment by amcasey** [comment]
@say25 The CI failures looked related to this change and I see there are some unresolved review comments.  There's no rush, but I wanted to let you know that the ball's in your court.

**Comment by say25** [comment]
> @say25 The CI failures looked related to this change and I see there are some unresolved review comments. There's no rush, but I wanted to let you know that the ball's in your court.

I think this is now building but it’s hard to tell what style comments I should address. Not sure if consistency is preferred here

**Comment by amcasey** [comment]
> > @say25 The CI failures looked related to this change and I see there are some unresolved review comments. There's no rush, but I wanted to let you know that the ball's in your court.
> 
> I think this is now building but it’s hard to tell what style comments I should address. Not sure if consistency is preferred here

Sorry, I'm not sure I understand your question.  Can you give an example of a style comment (I assume you mean warning) that would cause inconsistency if it were fixed?

File: src/Middleware/HttpOverrides/src/ForwardedHeaders.cs [hunk 3]
```diff
@@ -30,7 +30,11 @@ public enum ForwardedHeaders
     /// </summary>
     XForwardedPrefix = 1 << 3,
     /// <summary>
+    /// Process X-Forwarded-Port, which identifies the port used by the client.
+    /// </summary>
+    XForwardedPort = 1 << 4,
+    /// <summary>
     /// Process X-Forwarded-For, X-Forwarded-Host, X-Forwarded-Proto and X-Forwarded-Prefix.
```
Comment (say25) [inline src/Middleware/HttpOverrides/src/ForwardedHeaders.cs:37]:
```suggestion
    /// Process X-Forwarded-For, X-Forwarded-Host, X-Forwarded-Proto, X-Forwarded-Prefix and X-Forwarded-Port.
```

File: src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs [hunk 4]
```diff
@@ -167,14 +167,31 @@ public void ApplyForwarders(HttpContext context)
...
         if (_options.ForwardedHeaders.HasFlag(ForwardedHeaders.XForwardedPrefix))
         {
             checkPrefix = true;
             forwardedPrefix = requestHeaders.GetCommaSeparatedValues(_options.ForwardedPrefixHeaderName);
             if (_options.RequireHeaderSymmetry
                 && ((checkFor && forwardedFor!.Length != forwardedPrefix.Length)
                     || (checkProto && forwardedProto!.Length != forwardedPrefix.Length)
-                    || (checkHost && forwardedHost!.Length != forwardedPrefix.Length)))
+                    || (checkHost && forwardedHost!.Length != forwardedPrefix.Length)
+                    || (checkHost && forwardedPort!.Length != forwardedPrefix.Length)))
```
Comment (amcasey) [inline src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs:194]:
```suggestion
                    || (checkPort && forwardedPort!.Length != forwardedPrefix.Length)))
```

File: src/Middleware/HttpOverrides/src/PublicAPI.Shipped.txt [hunk 5]
```diff
@@ -15,9 +15,11 @@
 ~static Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersDefaults.XForwardedForHeaderName.get -> string
```
Comment (amcasey) [inline src/Middleware/HttpOverrides/src/PublicAPI.Shipped.txt:15]: Please revert the changes to this file - it's updated automatically.

File: src/Middleware/HttpOverrides/src/PublicAPI.Unshipped.txt [hunk 6]
```diff
@@ -5,8 +5,11 @@ Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.ForwardedPrefixHeaderName.g
 Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.ForwardedPrefixHeaderName.set -> void
```
Comment (amcasey) [inline src/Middleware/HttpOverrides/src/PublicAPI.Unshipped.txt:5]: I believe you want to add the following rows and revert everything else
```
Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.ForwardedPortHeaderName.get -> string!
Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.ForwardedPortHeaderName.set -> void
Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.OriginalPortHeaderName.get -> string!
Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.OriginalPortHeaderName.set -> void
Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedPort = 16 -> Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders
static Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersDefaults.XForwardedPortHeaderName.get -> string!
static Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersDefaults.XOriginalPortHeaderName.get -> string!
Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.All = Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedFor | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedHost | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedProto | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedPrefix | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedPort -> Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders
```

**Comment by say25** [comment]
Thanks for your help and patience with this @amcasey I did my best to address your comments. Sorry I'm so bad at this 😅

**Comment by amcasey** [comment]
Gotta start somewhere. :smile:  Thanks for putting in the time!

File: src/Middleware/HttpOverrides/src/PublicAPI.Unshipped.txt [hunk 7]
```diff
@@ -12,3 +12,11 @@ static Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersDefaults.XForwardedPre
...
+Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.ForwardedPortHeaderName.get -> string!
+Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.ForwardedPortHeaderName.set -> void
+Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.OriginalPortHeaderName.get -> string!
+Microsoft.AspNetCore.Builder.ForwardedHeadersOptions.OriginalPortHeaderName.set -> void
+Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedPort = 16 -> Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders
+static Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersDefaults.XForwardedPortHeaderName.get -> string!
+static Microsoft.AspNetCore.HttpOverrides.ForwardedHeadersDefaults.XOriginalPortHeaderName.get -> string!
+Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.All = Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedFor | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedHost | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedProto | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedPrefix | Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders.XForwardedPort -> Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders
```
Comment (amcasey) [inline src/Middleware/HttpOverrides/src/PublicAPI.Unshipped.txt:22]: My bad: I forgot to say that you'll need to remove the old new version of `All` on line 9 (but keep the `*REMOVED*` one).

File: src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs [hunk 8]
```diff
@@ -167,14 +167,31 @@ public void ApplyForwarders(HttpContext context)
             entryCount = Math.Max(forwardedHost.Length, entryCount);
         }
 
+        if (_options.ForwardedHeaders.HasFlag(ForwardedHeaders.XForwardedPort))
+        {
+            checkPort = true;
+            forwardedPort = requestHeaders.GetCommaSeparatedValues(_options.ForwardedPortHeaderName);
+            if (_options.RequireHeaderSymmetry
+                && ((checkFor && forwardedFor!.Length != forwardedPort.Length)
```
Comment (amcasey) [inline src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs:175]: I'll be honest and say that I don't know what header symmetry is (it appears to require some amount of similarity between parallel headers), but it's not obvious to me that this check is correct.  For one thing, the old pattern appears to be that the first block made 1 check, the second block made 2 checks, and the third block made 3 checks.  Now, for some reason, the third block makes 4 checks.

Comment (amcasey) [reply src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs:175]: Oh, the last check is comparing `forwardedPort!.Length` to itself - that's probably not necessary.  That, at least, gets us back to a linearly increasing number of checks.

File: src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs [hunk 9]
```diff
@@ -167,14 +167,31 @@ public void ApplyForwarders(HttpContext context)
...
             checkPrefix = true;
             forwardedPrefix = requestHeaders.GetCommaSeparatedValues(_options.ForwardedPrefixHeaderName);
             if (_options.RequireHeaderSymmetry
                 && ((checkFor && forwardedFor!.Length != forwardedPrefix.Length)
                     || (checkProto && forwardedProto!.Length != forwardedPrefix.Length)
-                    || (checkHost && forwardedHost!.Length != forwardedPrefix.Length)))
+                    || (checkHost && forwardedHost!.Length != forwardedPrefix.Length)
+                    || (checkPort && forwardedPort!.Length != forwardedPrefix.Length)))
             {
                 _logger.LogWarning(1, "Parameter count mismatch between X-Forwarded-Prefix and X-Forwarded-Host and X-Forwarded-For or X-Forwarded-Proto.");
```
Comment (amcasey) [inline src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs:196]: Seems like the log message probably needs to be updated.

File: src/Middleware/HttpOverrides/test/ForwardedHeadersMiddlewareTest.cs [hunk 10]
```diff
@@ -1241,15 +1241,16 @@ public async Task XForwardedPrefixInvalidPath(string forwardedPrefix)
     }
 
     [Theory]
-    [InlineData("11.111.111.11", "host1, host2", "h1, h2", "/prefix1, /prefix2")]
-    [InlineData("11.111.111.11, 22.222.222.22", "host1", "h1, h2", "/prefix1, /prefix2")]
-    [InlineData("11.111.111.11, 22.222.222.22", "host1, host2", "h1", "/prefix1, /prefix2")]
-    [InlineData("11.111.111.11, 22.222.222.22", "host1, host2", "h1, h2", "/prefix1")]
+    [InlineData("11.111.111.11", "host1, host2", "h1, h2", "/prefix1, /prefix2", "443")]
```
Comment (amcasey) [inline src/Middleware/HttpOverrides/test/ForwardedHeadersMiddlewareTest.cs:1244]: Strictly speaking, there's no particular reason to have a parameter if the argument is the same every time.  I'd probably vary the port from test to test though, in practice, it probably doesn't matter.

Perhaps it would be interesting to use null to mean "don't set that header" though?

**Comment by amcasey** [comment]
It should be uncontroversial, but we'll need an [API proposal](https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=) issue for this, since it affects the public API and, strictly speaking, breaks it by changing an enum value (which you can see from the API file, we've already done).

File: src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs (hunk 9 above)
Comment (say25) [reply src/Middleware/HttpOverrides/src/ForwardedHeadersMiddleware.cs:196]:
```suggestion
                _logger.LogWarning(1, "Parameter count mismatch between X-Forwarded-Prefix and X-Forwarded-Host and X-Forwarded-For and X-Forwarded-Proto or X-Forwarded-Port.");
```

**Comment by adityamandaleeka** [comment]
Looks like this PR has been sitting for a while with no updates. @say25 if you're still interested in working on this again at some point, feel free to open another PR but for now I'll close this.
